cmd/shim/main.go

```diff
@@ -148,6 +167,79 @@ func shim() int {
 	return exitCode
 }
 
+type SecretScrubWriter struct {
+	mu            sync.Mutex
+	w             io.Writer
+	secretToScrub map[string][]string
+	secretValues  map[string]string
+}
+
+func (w *SecretScrubWriter) Write(b []byte) (int, error) {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	s := string(b)
+	for _, id := range w.secretToScrub["env"] {
+		secret, ok := w.secretValues[id]
+		if !ok {
+			return -1, fmt.Errorf("secret value not available for: %v", id)
+		}
+		// FIXME: I think we can do better
+		s = strings.ReplaceAll(s, secret, "***")
```

Comment:
Leaving a note before I forget: I think we should be careful to handle writes of credentials that span multiple `Write([]byte)` calls. If I have a secret `foobar` and `Write([]byte("blahblahfoo"))` is followed by `Write([]byte("barblahblah")` it'd be great if we logged `blahblah***blahblah` (or whatever redaction syntax we choose).

We had to deal with this in Concourse, if you want to reference the code: [here](https://github.com/concourse/concourse/blob/da2a805c62eaba12437980655a372bf751dd74a4/atc/engine/log_event_writer.go#L68) is the writer and [here](https://github.com/concourse/concourse/blob/da2a805c62eaba12437980655a372bf751dd74a4/atc/engine/build_step_delegate.go#L453-L461) is where we do the actual scrubbing. Our implementation might look different though since we have different info at different times.

The basic idea we went with is to buffer writes line-by-line and scrub secrets line-by-line. So if you print a multi-line secret you'll get multiple scrubbed lines. The downside is logs get buffered linewise, so it'd be nice if we only did the scrubbing if any credentials were actually configured. Otherwise test frameworks that print `..................` will look unresponsive.

Reply:
Thanks, I was thinking about that actually, but I wanted to reserve that for a consequent PR.
The current implementation of the log scrubber from 0.2 also had this limitation. So at least, if we're in par with 0.2, we're better than today.
But yeah, a secret that spans many `Write()` could be missed without a real good implementation with some buffer. I'll have a look at your code, thank you!

Also, @grouville sent me this: https://github.com/dagger/dagger/issues/1892